torch/_inductor/codegen/triton.py

```diff
@@ -197,57 +260,73 @@ def create(
-        reshape_suffix = [f"{t.prefix.upper()}BLOCK" for t in range_trees]
+        final_shape = [get_block_size(tree) for tree in range_trees]
+
+        sizevars = V.graph.sizevars
+
+        def lookup_size(exprs: Iterable[sympy.Expr]) -> List[sympy.Expr]:
+            return [sizevars.lookup_precomputed_size(expr) for expr in exprs]
 
-        # Only drop broadcast dims if the output has the same
-        # rank as the block. Otherwise, we will get shape errors.
-        drop_broadcasts = len(reshape_suffix) == len(params.strides)
+        # Look up precomputed sizes
+        params.shape = lookup_size(params.shape)
+        params.strides = lookup_size(params.strides)
 
-        broadcasting_dim = [s == 0 for s in params.strides]
-        for i, is_broadcasting in enumerate(broadcasting_dim):
-            if is_broadcasting and drop_broadcasts:
-                # drop any stride==0 dimensions for performance
-                reshape_suffix[i] = "1"
+        # Strip out dimensions of stride 0.
+        # These will be restored with tl.broadcast_to.
+        broadcasting_dims = [
+            sizevars.statically_known_equals(stride, 0) for stride in params.strides
+        ]
+
+        # Strip out dimensions of size 1.
+        # These will be restored by tl.reshape.
+        singleton_dims = [
+            sizevars.statically_known_equals(dim, 1) for dim in params.block_shape
+        ]
+        if all(singleton_dims):
+            # Handle a pure singletons, e.g. [1, 1]
+            singleton_dims[-1] = False
 
         if V.kernel.no_x_dim:
             assert range_trees[0].prefix == "x"
-            reshape_suffix.pop(0)
+            final_shape.pop(0)
 
         if (
             not V.kernel.inside_reduction
             and len(params.strides) == len(V.kernel.numels) - 1
             and V.kernel.numels[-1] != 1
         ):
             # Need to expand rank by 1 to match rank when self.inside_reduction=True
-            reshape_suffix.append("1")
+            final_shape.append(sympy.Integer(1))
 
-        def filter(it):
-            """Removes any broadcasting dims from a given sequence"""
-            assert len(it) == len(broadcasting_dim)
+        def remove_dims(it):
+            """Removes any broadcasting or singleton dims from a given sequence"""
+            assert len(it) == len(broadcasting_dims)
             return [
                 item
-                for item, is_broadcasting in zip(it, broadcasting_dim)
-                if not is_broadcasting or not drop_broadcasts
+                for item, is_broadcasting, is_singleton in zip(
+                    it, broadcasting_dims, singleton_dims
+                )
+                if not (is_broadcasting or is_singleton)
             ]
 
         # Drop broadcasting dimensions from the input.
+        broadcast_shape = [
+            dim
+            for dim, is_singleton in zip(params.block_shape, singleton_dims)
+            if not is_singleton
```

Comment:
need record dimensions in `is_broadcasting` as well?

Reply:
The idea here is that `broadcast_shape` stores the shape after broadcasting. The pre-broadcast shape is the same as this, except we replace dims with `is_broadcasting=True` with 1. (See line 224.) I can add a comment to clarify.

Reply:
I cleaned up the code a bit, and added a comment to clarify this in [72dabc7](https://github.com/pytorch/pytorch/pull/135557/commits/72dabc7fc9c84e21d2eba107b4cb04fe47f065b0)